src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj

```diff
@@ -20,55 +17,42 @@
                           jruby/JRubyPuppetService
                           [[:ConfigService get-config]
                            [:ShutdownService shutdown-on-error]
-                           [:PuppetProfilerService get-profiler]]
+                           [:PuppetProfilerService get-profiler]
+                           [:PoolManagerService create-pool]]
   (init
     [this context]
-    (let [config            (core/initialize-config (get-config))
-          service-id        (tk-services/service-id this)
-          agent-shutdown-fn (partial shutdown-on-error service-id)
-          profiler          (get-profiler)]
-      (core/verify-config-found! config)
-      (log/info "Initializing the JRuby service")
-      (if (:use-legacy-auth-conf config)
-        (log/warn "The 'jruby-puppet.use-legacy-auth-conf' setting is set to"
-                  "'true'.  Support for the legacy Puppet auth.conf file is"
-                  "deprecated and will be removed in a future release.  Change"
-                  "this setting to 'false' and migrate your authorization rule"
-                  "definitions in the /etc/puppetlabs/puppet/auth.conf file to"
-                  "the /etc/puppetlabs/puppetserver/conf.d/auth.conf file."))
-      (core/add-facter-jar-to-system-classloader (:ruby-load-path config))
-      (let [pool-context (core/create-pool-context config profiler agent-shutdown-fn)]
-        (jruby-agents/send-prime-pool! pool-context)
-        (-> context
-            (assoc :pool-context pool-context)
-            (assoc :borrow-timeout (:borrow-timeout config))
-            (assoc :event-callbacks (atom []))
-            (assoc :environment-class-info-tags (atom {}))))))
+    (let [jruby-config (core/initialize-and-create-jruby-config
+                        (get-config)
+                        (get-profiler)
+                        (partial shutdown-on-error (tk-services/service-id this)))
+          pool-context (create-pool jruby-config)]
+      (-> context
+          (assoc :pool-context pool-context)
+          (assoc :environment-class-info-tags (atom {})))))
+
   (stop
    [this context]
-   (let [{:keys [pool-context]} (tk-services/service-context this)
-         on-complete (promise)]
-     (log/debug "Beginning flush of JRuby pools for shutdown")
-     (jruby-agents/send-flush-pool-for-shutdown! pool-context on-complete)
-     @on-complete
-     (log/debug "Finished flush of JRuby pools for shutdown"))
+   (let [{:keys [pool-context]} (tk-services/service-context this)]
+     (jruby-core/flush-pool-for-shutdown! pool-context))
    context)
 
   (borrow-instance
     [this reason]
-    (let [{:keys [pool-context borrow-timeout event-callbacks]} (tk-services/service-context this)]
-      (core/borrow-from-pool-with-timeout pool-context borrow-timeout reason @event-callbacks)))
+    (let [{:keys [pool-context]} (tk-services/service-context this)
```

Comment:
Based on the current state of the code, it appeared to me that `borrow-instance` and `return-instance` at the service level are now only called by test code - since production code only ever calls `with-jruby-instance` (preceded by the new `get-pool-context` function).  If these are only called by test code, I'd be inclined to just make these functions in the jruby-testutils code and have those helpers use `get-pool-context` and `with-jruby-instance`, as appropriate.  WDYT?